File: tigerml/lib/calc.ml
let example_prog = "3 * 5 + 4 * 6\n" (* 39 *)

let main () =
  Printf.printf "Running calc.ml\n";
  Printf.printf "Program: %s\n" example_prog;
  try
    let lexbuf = Lexing.from_string example_prog in
    while true do
      let result = Parser.main Lexer.token lexbuf in
      Printf.printf "Result: %s\n" (Ast_calc.to_string result)
    done
    (* on error, return *)
  with Lexer.Eof -> ()

File: tigerrs/src/calc.rs
use crate::lexer::Token;
use logos::Logos;

use lalrpop_util::lalrpop_mod;

lalrpop_mod!(pub grammar);

pub fn main() {
    let example_prog = "3 * 5 + 4 * 6";
    println!("Running calc.rs");
    println!("Program: {}", example_prog);

    let lexer = Token::lexer(example_prog)
        .spanned()
        .map(|(tok, span)| match tok {
            Ok(Token::Error) => Err("Lexer error".to_string()),
            Err(_) => Err("Parse error".to_string()),
            Ok(t) => Ok((span.start, t, span.end)),
        });

    let parser = grammar::ExprParser::new();

    match parser.parse(lexer) {
        Ok(result) => println!("Result: {}", result),
        Err(e) => println!("Error: {:?}", e),
    }
}
